File: src/sp/it/pl/gui/objects/picker/Picker.java
package sp.it.pl.gui.objects.picker;

import java.util.List;
import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Stream;
import javafx.css.PseudoClass;
import javafx.geometry.Insets;
import javafx.geometry.VPos;
import javafx.scene.Node;
import javafx.scene.control.Label;
import javafx.scene.control.ScrollPane;
import javafx.scene.layout.Pane;
import javafx.scene.layout.Region;
import javafx.scene.layout.StackPane;
import sp.it.pl.gui.objects.Text;
import sp.it.pl.util.animation.Anim;
import sp.it.pl.util.functional.Functors.Ƒ1;
import static java.lang.Math.ceil;
import static java.lang.Math.floor;
import static java.lang.Math.max;
import static java.lang.Math.sqrt;
import static java.util.Arrays.asList;
import static javafx.scene.control.ScrollPane.ScrollBarPolicy.NEVER;
import static javafx.scene.input.MouseButton.PRIMARY;
import static javafx.scene.input.MouseButton.SECONDARY;
import static javafx.scene.layout.Region.USE_COMPUTED_SIZE;
import static javafx.scene.text.TextAlignment.JUSTIFY;
import static javafx.util.Duration.millis;
import static sp.it.pl.util.animation.Anim.animPar;
import static sp.it.pl.util.animation.Anim.animSeq;
import static sp.it.pl.util.functional.Util.byNC;
import static sp.it.pl.util.functional.Util.filter;
import static sp.it.pl.util.functional.Util.forEachWithI;
import static sp.it.pl.util.graphics.Util.layScrollVText;
import static sp.it.pl.util.graphics.UtilKt.setScaleXY;

/**
 * Generic item picker.
 * <p/>
 * Node displaying elements as grid.
 * Elements are converted to their text representation according to provided
 * mapper. Element should override toString() method if no mapper is provided.
 * <p/>
 * Elements will be sorted lexicographically.
 * <p/>
 * Elements will be represented graphically depending on the cell factory.
 */
public class Picker<E> {

	/** Style class for cell. */
	public static final String STYLE_CLASS = "item-picker";
	public static final List<String> CELL_STYLE_CLASS = asList("block", "item-picker-element");

	private final CellPane tiles = new CellPane();
	public final ScrollPane root = new ScrollPane(tiles);
	public boolean consumeCancelClick = false;
	/**
	 * Nonnull procedure executed when item is selected passing the item as parameter.
	 * Default implementation does nothing.
	 */
	public Consumer<? super E> onSelect = item -> {};
	/**
	 * Nonnull procedure executed when no item is selected. Invoked when user cancels
	 * the picking by right click.
	 * Default implementation does nothing.
	 * <p/>
	 * For example one might want to close this picker when no item is selected.
	 */
	public Runnable onCancel = () -> {};
	/**
	 * Nonnull text factory.
	 * Creates string name/title of the item.
	 * Default implementation calls {@link Objects#toString(Object)}
	 */
	public Function<? super E, ? extends String> textConverter = Objects::toString;
	/**
	 * Nonnull info text factory.
	 * Creates string descriptive of the item.
	 * Default implementation returns empty string.
	 */
	public Function<? super E, ? extends String> infoConverter = item -> "";
	/** Nonnull item supplier. Fetches the items as a stream. Default implementation returns empty stream. */
	public Supplier<? extends Stream<? extends E>> itemSupply = Stream::empty;
	/**
	 * Nonnull cell factory.
	 * Creates graphic representation of the item.
	 * Also might define minimum and maximum item size.
	 */
	public Ƒ1<E,Pane> cellFactory = item -> {
		String text = textConverter.apply(item);
		Label l = new Label(text);
		StackPane cell = new StackPane(l);
		cell.setMinSize(90, 30);
		cell.getStyleClass().setAll(CELL_STYLE_CLASS);

		// set up info pane
		String info = infoConverter.apply(item);
		if (!info.isEmpty()) {
			// info content
			Node content = cell.getChildren().get(0);
			Text ta = new Text(info);
			ta.setTextOrigin(VPos.CENTER);
			ta.setMouseTransparent(true);
			ta.setTextAlignment(JUSTIFY);
			ScrollPane sp = layScrollVText(ta);
			cell.getChildren().add(sp);
			cell.setPadding(new Insets(20));
			// animation
			Anim anim = new Anim(millis(300), x -> {
				sp.setOpacity(x);
				content.setOpacity(1 - x*x);
				setScaleXY(sp, 0.7 + 0.3*x*x);
			}).applyNow();
			cell.hoverProperty().addListener((o, ov, nv) -> anim.playFromDir(nv));
		}

		return cell;
	};

	public Picker() {
		root.setMinSize(USE_COMPUTED_SIZE, USE_COMPUTED_SIZE);
		root.setPrefSize(USE_COMPUTED_SIZE, USE_COMPUTED_SIZE);  // leave resizable
		root.setMaxSize(USE_COMPUTED_SIZE, USE_COMPUTED_SIZE);
		root.setPannable(false);  // forbid mouse panning (can cause unwanted horizontal scrolling)
		root.setFitToWidth(true); // make content horizontally resize with scroll pane
		root.setHbarPolicy(NEVER);  // thus no need for horizontal scrollbar
		root.setOnMouseClicked(e -> {
			// right click runs cancel, sometimes its important to consume the event, sometimes
			// it is important for it to pass through. so I left it configurable until I find a
			// better way
			if (e.getButton()==SECONDARY) {
				onCancel.run();
				if (consumeCancelClick) e.consume();
			}
		});
		root.getStyleClass().add(STYLE_CLASS);
	}

	public void buildContent() {
		tiles.getChildren().clear();
		// get items
		itemSupply.get()
			// & sort
			.sorted(byNC(textConverter::apply))
			// & create cells
			.forEach(item -> {
				Node cell = cellFactory.apply(item);
				cell.setOnMouseClicked(e -> {
					if (e.getButton()==PRIMARY) {
						onSelect.accept(item);
						e.consume();
					}
				});
				tiles.getChildren().add(cell);
			});

		Pane p = new Pane();
		p.getStyleClass().setAll(CELL_STYLE_CLASS);
		p.pseudoClassStateChanged(PseudoClass.getPseudoClass("empty"), true);
		p.getProperties().put("empty_cell", null);
		p.setManaged(false);
		tiles.getChildren().add(p);

		// animate & show
		int s = tiles.getChildren().size();
		animPar(tiles.getChildren(), (i, n) -> animSeq(
			new Anim(n::setOpacity).dur(millis(i*(750/s))).intpl(x -> 0.0),
			new Anim(n::setOpacity).dur(millis(500)).intpl(x -> sqrt(x))
		)).play();
	}

	public Region getNode() {
		buildContent();
		return root;
	}

	@SuppressWarnings("unchecked")
	private List<Region> getCells() {
		return (List) filter(tiles.getChildren(), it -> !it.getProperties().containsKey("empty_cell"));
	}

	private class CellPane extends Pane {

		@Override
		protected void layoutChildren() {
			Insets padding = root.getPadding();
			double width = root.getWidth() - padding.getLeft() - padding.getRight();
			double height = root.getHeight() - padding.getTop() - padding.getBottom();

			int gap = 1;
			int elements = getCells().size();
			double min_cell_w = max(1, getCells().get(0).getMinWidth());
			double min_cell_h = max(1, getCells().get(0).getMinHeight());
			// if (elements==0) return;

			int c = width>height ? (int) ceil(sqrt(elements)) : (int) floor(sqrt(elements));
			c = width<c*min_cell_w ? (int) floor(width/min_cell_w) : c;
			final int columns = max(1, c);

			int rows = (int) ceil(elements/(double) columns);

			double sumgapy = (rows - 1)*gap;
			final double cell_height = height<rows*min_cell_h ? min_cell_h : (height - sumgapy)/rows - 1/(double) rows;

			double W = rows*(cell_height + gap) - gap>height ? width - 15 : width; // take care of scrollbar
			double sumgapx = (columns - 1)*gap;  // n elements have n-1 gaps
			final double cell_width = (W - sumgapx)/columns;

			int emptyCells = columns*rows-getCells().size();
			forEachWithI(getCells(), (i, n) -> {
				double x = padding.getLeft() + i%columns*(cell_width + gap);
				double y = padding.getTop() + i/columns*(cell_height + gap);
				n.resizeRelocate(
					snapPositionX(x),
					snapPositionY(y),
					snapPositionX(x+cell_width)-snapPositionX(x),
					snapPositionY(y+cell_height)-snapPositionY(y)
				);
			});

			Node emptyCell = getChildren().stream().filter(it -> it.getProperties().containsKey("empty_cell")).findFirst().get();
			if (emptyCells!=0) {
				int emptyCellI = getCells().size();
				double x = padding.getLeft() + emptyCellI%columns*(cell_width + gap);
				double y = padding.getTop() + emptyCellI/columns*(cell_height + gap);

				emptyCell.resizeRelocate(
					snapPositionX(x),
					snapPositionY(y),
					snapPositionX(root.getWidth()-padding.getRight())-snapPositionX(x),
					snapPositionY(y+cell_height)-snapPositionY(y)
				);
			} else {
				emptyCell.resize(0.0, 0.0);
			}
		}

	}
}

File: src/sp/it/pl/gui/objects/picker/Picker.kt
package sp.it.pl.gui.objects.picker

import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.geometry.VPos.CENTER
import javafx.scene.control.ScrollPane
import javafx.scene.control.ScrollPane.ScrollBarPolicy.NEVER
import javafx.scene.input.MouseButton.PRIMARY
import javafx.scene.input.MouseButton.SECONDARY
import javafx.scene.layout.Pane
import javafx.scene.layout.Region
import javafx.scene.layout.Region.USE_COMPUTED_SIZE
import javafx.scene.text.TextAlignment
import sp.it.pl.util.animation.Anim.Companion.anim
import sp.it.pl.util.animation.Anim.Companion.animPar
import sp.it.pl.util.animation.Anim.Companion.animSeq
import sp.it.pl.util.functional.setTo
import sp.it.pl.util.graphics.label
import sp.it.pl.util.graphics.lay
import sp.it.pl.util.graphics.pane
import sp.it.pl.util.graphics.pseudoclass
import sp.it.pl.util.graphics.scrollText
import sp.it.pl.util.graphics.setMinPrefMaxSize
import sp.it.pl.util.graphics.setScaleXY
import sp.it.pl.util.graphics.stackPane
import sp.it.pl.util.graphics.text
import sp.it.pl.util.math.div
import sp.it.pl.util.math.max
import sp.it.pl.util.math.millis
import sp.it.pl.util.math.times
import sp.it.pl.util.reactive.attach
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.sqrt

/**
 * Generic picker.
 *
 * The elements are obtained using [itemSupply] when [buildContent] is called, sorted lexicographically and disaplayed
 * in a 2D grid using [textConverter], [infoConverter].
 */
open class Picker<E> {

    private val tiles = CellPane()
    /** Scene graph root of this object. */
    @JvmField public val root = ScrollPane(tiles)
    /** Invoked when item is selected. Default implementation does nothing. */
    @JvmField public var onSelect: (E) -> Unit = {}
    /** Invoked when user user cancels the picking. Default implementation does nothing. */
    @JvmField public var onCancel: () -> Unit = {}
    /** It may be desirable to consume the mouse click event that caused the cancellation. Default false. */
    @JvmField public var consumeCancelEvent = false
    /** Cell text factory producing name/title of the item. Default implementation calls [Any.toString] */
    @JvmField public var textConverter: (E) -> String = Any?::toString
    /** Cell detail text text factory producing description of the item. Default implementation returns empty string. */
    @JvmField public var infoConverter: (E) -> String = { "" }
    /** Supplier that returns items to be displayed. Default implementation returns empty sequence. */
    @JvmField public var itemSupply: () -> Sequence<E> = { sequenceOf() }

    private val cellFactory: (E) -> Pane = { item ->
        stackPane() {
            setMinSize(90.0, 30.0)
            styleClass += CELL_STYLE_CLASS
            padding = Insets(20.0)

            val contentText = textConverter(item)
            val contentInfoText = infoConverter(item)
            val content = label(contentText)

            lay += content

            if (contentInfoText.isNotEmpty()) {
                val contentInfo = scrollText {
                    text(contentText) {
                        textOrigin = CENTER
                        isMouseTransparent = true
                        textAlignment = TextAlignment.CENTER
                    }
                }
                lay += contentInfo

                val anim = anim(300.millis) {
                    content.opacity = 1-it*it
                    contentInfo.opacity = it
                    contentInfo.setScaleXY(0.7+0.3*it*it)
                }

                anim.applyNow()
                hoverProperty() attach { anim.playFromDir(it) }
            }
        }
    }

    public constructor() {
        root.apply {
            setMinPrefMaxSize(USE_COMPUTED_SIZE)
            isPannable = false  // forbid mouse panning (can cause unwanted horizontal scrolling)
            isFitToWidth = true // make content horizontally resize with scroll pane
            hbarPolicy = NEVER
            setOnMouseClicked(EventHandler {
                if (it.button==SECONDARY) {
                    onCancel()
                    if (consumeCancelEvent) it.consume()
                }
            })
            styleClass += STYLE_CLASS
        }
    }

    public fun buildContent() {
        tiles.children setTo itemSupply()
                .sortedBy(textConverter)
                .map { item ->
                    cellFactory(item).apply {
                        onMouseClicked = EventHandler {
                            if (it.button==PRIMARY) {
                                onSelect(item)
                                it.consume()
                            }
                        }
                    }
                }
        tiles.children += pane {
            styleClass += CELL_STYLE_CLASS
            pseudoClassStateChanged(pseudoclass("empty"), true)
            properties[KEY_EMPTY_CELL] = null
            isManaged = false
        }

        // animate & show
        animPar(tiles.children) { i, n ->
            animSeq(
                    anim(750.millis*i/tiles.children.size) { n.opacity = it }.intpl { 0.0 },
                    anim(500.millis) { n.opacity = it }.intpl { sqrt(it) }
            )
        }.play()
    }

    @Suppress("UNCHECKED_CAST")
    private fun getCells() = tiles.children.filter { !it.properties.containsKey(KEY_EMPTY_CELL) } as List<Region>

    private inner class CellPane: Pane() {

        override protected fun layoutChildren() {
            val cells = getCells()
            val padding = root.getPadding()
            val gap = 1.0
            val width = root.width-padding.left-padding.right
            val height = root.height-padding.top-padding.bottom

            val elements = cells.size
            val cellMinWidth = 1.0 max cells.first().minWidth
            val cellMinHeight = 1.0 max cells.first().minHeight

            var c = if (width>height) ceil(sqrt(elements.toDouble())).toInt() else floor(sqrt(elements.toDouble())).toInt()
            c = if (width<c*cellMinWidth) floor(width/cellMinWidth).toInt() else c
            val columns = max(1, c)

            val rows = ceil(elements.toDouble()/columns).toInt()

            val gapSumY = (rows-1)*gap
            val cellHeight = if (height<rows*cellMinHeight) cellMinHeight else (height-gapSumY)/rows-1.0/rows

            val W = if (rows*(cellHeight+gap)-gap>height) width-15 else width // TODO: take care of scrollbar better
            val gapSumX = (columns-1)*gap
            val cellWidth = (W-gapSumX)/columns

            cells.forEachIndexed { i, n ->
                val x = padding.left+i%columns*(cellWidth+gap)
                val y = padding.top+i/columns*(cellHeight+gap)
                n.resizeRelocate(
                        x.snapX,
                        y.snapY,
                        (x+cellWidth).snapX-x.snapX,
                        (y+cellHeight).snapY-y.snapY
                )
            }

            val needsEmptyCell = 0==columns*rows-cells.size
            val emptyCell = children.find { it.properties.containsKey(KEY_EMPTY_CELL) }!!
            if (needsEmptyCell) {
                val i = cells.size
                val x = padding.left+i%columns*(cellWidth+gap)
                val y = padding.top+i/columns*(cellHeight+gap)

                emptyCell.resizeRelocate(
                        x.snapX,
                        y.snapY,
                        (root.width-padding.right).snapX-x.snapX,
                        (y+cellHeight).snapY-y.snapY
                )
            } else {
                emptyCell.resize(0.0, 0.0)
            }
        }

        private val Double.snapX: Double get() = snapPositionX(this)
        private val Double.snapY: Double get() = snapPositionY(this)
    }

    companion object {
        val STYLE_CLASS = "item-picker"
        val CELL_STYLE_CLASS = listOf("block", "item-picker-element")
        private const val KEY_EMPTY_CELL = "empty_cell"
    }

}